File: openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/utils/Hashes.java
package io.lacuna.artifex.utils;

/**
 * @author ztellman
 */
public class Hashes {

  public static int hash(double x) {
    long hash = 31L * Double.doubleToLongBits(x);
    return (int) (hash ^ (hash >>> 32));
  }

  public static int hash(double x, double y) {
    long hash = 1L;
    hash = (hash * 31) + Double.doubleToLongBits(x);
    hash = (hash * 31) + Double.doubleToLongBits(y);
    return (int) (hash ^ (hash >>> 32));
  }

  public static int hash(double x, double y, double z) {
    long hash = 1L;
    hash = (hash * 31) + Double.doubleToLongBits(x);
    hash = (hash * 31) + Double.doubleToLongBits(y);
    hash = (hash * 31) + Double.doubleToLongBits(z);
    return (int) (hash ^ (hash >>> 32));
  }

  public static int hash(double x, double y, double z, double w) {
    long hash = 1L;
    hash = (hash * 31) + Double.doubleToLongBits(x);
    hash = (hash * 31) + Double.doubleToLongBits(y);
    hash = (hash * 31) + Double.doubleToLongBits(z);
    hash = (hash * 31) + Double.doubleToLongBits(w);
    return (int) (hash ^ (hash >>> 32));
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/utils/Hashes.kt
package org.openrndr.utils

/**
 * @author ztellman
 */
object Hashes {
    fun hash(x: Double): Int {
        val hash: Long = 31L * x.toBits()
        return (hash xor (hash ushr 32)).toInt()
    }

    fun hash(x: Double, y: Double): Int {
        var hash = 1L
        hash = hash * 31 + x.toBits()
        hash = hash * 31 + y.toBits()
        return (hash xor (hash ushr 32)).toInt()
    }

    fun hash(x: Double, y: Double, z: Double): Int {
        var hash = 1L
        hash = hash * 31 + x.toBits()
        hash = hash * 31 + y.toBits()
        hash = hash * 31 + z.toBits()
        return (hash xor (hash ushr 32)).toInt()
    }

    fun hash(x: Double, y: Double, z: Double, w: Double): Int {
        var hash = 1L
        hash = hash * 31 + x.toBits()
        hash = hash * 31 + y.toBits()
        hash = hash * 31 + z.toBits()
        hash = hash * 31 + w.toBits()
        return (hash xor (hash ushr 32)).toInt()
    }
}